kernel/timeout.c

```diff
@@ -44,37 +41,25 @@ static struct _timeout *next(struct _timeout *t)
 	return n == NULL ? NULL : CONTAINER_OF(n, struct _timeout, node);
 }
 
-static void remove_timeout(struct _timeout *t)
-{
-	if (t->node.next != NULL && t->node.prev != NULL) {
-		if (next(t) != NULL) {
-			next(t)->dticks += t->dticks;
-		}
-
-		sys_dlist_remove(&t->node);
-	}
-	t->node.next = t->node.prev = NULL;
-	t->dticks = _INACTIVE;
-}
-
-static s32_t elapsed(void)
-{
-	return announce_remaining == 0 ? z_clock_elapsed() : 0;
-}
-
 void _add_timeout(struct _timeout *to, _timeout_func_t fn, s32_t ticks)
 {
-	__ASSERT(to->dticks < 0, "");
+	__ASSERT(!sys_dnode_is_linked(&to->node), "");
 	to->fn = fn;
+
+	/* @todo This really ought to be removed to allow scheduling
+	 * with negative delays, since the floor operation fails to
+	 * maintain correct periodicity for timers that are so late
+	 * they missed more than an interval.  But without it
+	 * kernel/timer/timer_api:test_timer_periodicity fails.  Is
+	 * the test making incorrect assumptions about how to trick
+	 * the system? */
```

Comment:
This is the point where I need insight.  Negative delays are important for maintaining periodicity when ticks are not announced in a timely manner.

Reply:
@pabigot: Question is what is reference point of `_add_timeout()`? As far as I know, the reference is "now()" in real time. So scheduling timeouts in past is not possible. On the other hand we should be able to add timeouts which should fire "now()".

Reply:
The incoming argument `ticks` is signed, relative to now = `k_uptime_get()`.  It's negative if we've already missed the deadline.  We can't unilaterally discard the only information we have on the alignment of the recurring deadline, so negative delays must be recorded and preserved so subsequent reschedules maintain the correct offsets.  In practice the timeout will be near the head of the queue (before anything that isn't late) so it will be invoked in the next tick interrupt.  That's the earliest possible unless the `z_clock_announce` is reworked to loop as long as there's a late timeout at the head of the list.

Reply:
Ok. Now i see it. We have to handle missed periodic timeouts which were missed more than the period.